DataViews: add support for infinite scroll in table, grid, and list layouts

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

It's likely this will be needed as DataViews is adopted in more wp-admin screens.

This PR adds an optional `infiniteScrollHandler` to the DataViews `pagination` object. If this handler exists, an infinite scroll toggle is enabled in the view config. 

If both `infiniteScrollHandler` exists and infinite scroll is toggled on in the view config, a scroll event handler is attached to the dataviews wrapper.

The PR also updates storybook with an infinite scroll story, though it currently has a double scrollbar as I had to add a fixed height container around it to trigger the scroll event. This won't happen in regular usage, at least provided the dataviews container isn't directly inside an iframe 😅 

To try this out, this PR also changes the PostList to pass an `infiniteScrollHandler` into DataViews, so its functionality can be tested on the Pages screen (provided there are enough pages to trigger pagination).

The PostList change **doesn't need to be done in this PR**, I just added it so it's easier to test. 


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Go to site editor > Pages;
2. You'll need to have at least 11 pages to trigger pagination (and set per page to 10)
3. Select Grid view in the layout picker
4. Open the Appearance dropdown and check there is an "infinite scroll" toggle (it's very basic for now, label should probably follow the styling of the other controls)
5. Enable infinite scroll and scroll down the page to see it work.
6. run `npm run storybook:dev` and check the infinite scroll story.


https://github.com/user-attachments/assets/e001a4d6-7dd2-42cd-9b29-9ceb063835c2


https://github.com/user-attachments/assets/065ae651-320b-4916-ab72-67159f9f92ce

tellthemachines committed:
Try adding optional infinite scroll to dataviews

tellthemachines commented:
> Any reason to keep this for grid for now, this seems generic to me, just like the pagination component that we have

No particular reason apart from it being a more common UI pattern in that type of view, but given it's an opt-in it should be fine to add it for all views.

tellthemachines committed:
try infinite scroll on all view types

tellthemachines committed:
make `infiniteScrollHandler` a property of `paginationInfo`

tellthemachines commented:
Thanks for the feedback Riad!

I've updated to add infiniteScrollHandler inside paginationInfo and allow infinite scroll on all view types. There's still a few things that need to be added before the PR is ready; I'll continue working on it in the next few days.

talldan commented:
Good question. I found this gallery which is creative commons - https://www.flickr.com/photos/lightsinthedark/albums/72157694956141104/

But I'm not sure how you get the links.

tellthemachines committed:
Add some more moons to storybook fixtures

tellthemachines committed:
Add infinite scroll story

tellthemachines committed:
update tests referring to data fixture

tellthemachines committed:
Add aria roles for feed pattern

tellthemachines commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -351,8 +362,13 @@ function ViewGrid< Item >( {
+											isInfiniteScroll
+												? 'feed'
+												: undefined
+										}
```
I'm really not sure if the semantics here are correct: when `groupByField` is enabled, this means there will be multiple `feed` role groups on the page. Should we just disable infinite scroll when grouping by field?

andrewserong replied:
Good question. I suppose semantically it'd be the containing `VStack` that would get the "feed" role if there should just be the one?

> Should we just disable infinite scroll when grouping by field?

From my perspective it would make sense to start out not supporting "group by" for now, and return to adding it further down the track if it's needed? Aside from the semantics of the `feed` role, it does feel like the group by case is a slightly different UX from scrolling through a single list 🤔

tellthemachines committed:
add feed aria roles to list and table layouts

tellthemachines committed:
try toggle group control

tellthemachines commented:
I had a look at [our recommendations](https://make.wordpress.org/accessibility/handbook/markup/design-patterns/infinite-scroll/) regarding the accessibility of the infinite scroll pattern. One thing to bear in mind is that the event handler that actually implements the infinite scroll by loading in more data, etc, is created by the consumer, so there's nothing that can be done on the dataviews side to safeguard a good navigation experience or memory management. 

The example implementation in edit-site post list is very simplified and doesn't deal with these matters, mostly because that view doesn't handle URL changes for pagination either, and adding proper navigation would require that. This is why I suggest we don't merge that change in this PR, and follow up with a proper implementation later if/when we want to add infinite scroll to that view.

tellthemachines committed:
Changelog entry

oandregal commented on packages/dataviews/src/components/dataviews-view-config/infinite-scroll-toggle.tsx:
```diff
@@ -0,0 +1,56 @@
+			value={ infiniteScrollEnabled ? 'enabled' : 'disabled' }
+			onChange={ ( value ) => {
+				const newValue = value === 'enabled';
+				onChangeView( {
```
When switching it off, we also need to set the page we are in and make sure we only list the perPage number of items (not all of them)? Otherwise, it's like nothing happened:

https://github.com/user-attachments/assets/b287dba2-8226-4b15-92b6-c0d02fa83af8

tellthemachines replied:
That was an issue with the storybook implementation, not the toggle control 😅 
It should be fixed in [8804ac9](https://github.com/WordPress/gutenberg/pull/70955/commits/8804ac9615c417120d90eca25e5daa5b67d285d4).

oandregal reviewed (commented):
I'm not sure if this is an artifact of the storybook, but I've noticed that the footer is not visible until you've scrolled down all the pages. This is problematic as you can't access bulk selection (in grid) and bulk actions (in all layouts).

https://github.com/user-attachments/assets/6ec44c19-da0c-42ed-a684-754be30b3c4c

oandregal commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -383,3 +385,119 @@ export const GroupedGridLayout = () => {
+				fields={ fields }
+				onChangeView={ setView }
+				actions={ actions }
+				isLoading={ isLoadingMore }
```
I wanted to try the "loading" experience. Layouts need to combine it with "has data":

- when there is no data, display a spinner instead like they do (e.g., [table](https://github.com/WordPress/gutenberg/blob/7134ceb1f0a5192d58dac43d9cc21cdd947b4548/packages/dataviews/src/dataviews-layouts/table/index.tsx#L495))
- when there is actually data but is still loading: what do we do? display a "loading item" at the end of the list?

https://github.com/user-attachments/assets/419e5733-ef30-4f9e-857b-4736880ce9ee

oandregal replied:
Additionally, what do we do about the existing pagination? The current state is that we display _the last page_ that has been fetched. And interacting with it doesn't do anything. Perhaps we hide it if infinite scroll is on?

https://github.com/user-attachments/assets/e3a08e6e-7abd-4ade-a72f-ff14da85b8e4

ramonjd replied:
> And interacting with it doesn't do anything. Perhaps we hide it if infinite scroll is on?

Interacting does something, but I don't think it's expected. The page jumps when you select pages after all have loaded, and the `n items` changes. 

https://github.com/user-attachments/assets/7bdfa252-a1e9-47da-bae6-180bdc5ee2ac

I wonder if infinite scroll needs some sort of custom footer, e.g., how many have been loaded and/or selected.

Otherwise it's working nicely for me.

tellthemachines replied:
> > Additionally, what do we do about the existing pagination? The current state is that we display the last page that has been fetched. And interacting with it doesn't do anything. Perhaps we hide it if infinite scroll is on?

Yeah I think we just hide the pagination bit.

>  how many have been loaded and/or selected

This bit is working fine so we can leave it as is for now.

>  when there is no data, display a spinner instead like they do (e.g., [table](https://github.com/WordPress/gutenberg/blob/7134ceb1f0a5192d58dac43d9cc21cdd947b4548/packages/dataviews/src/dataviews-layouts/table/index.tsx#L495))

When would we have no data? If the server doesn't respond, shouldn't we display an error message in that case?

tellthemachines replied:
Ok I think I get it. When the data is slow to load on the page, the spinner displays. That'll happen for the initial data whether infinite scroll is enabled or not. 

If we're loading additional data with infinite scroll, that's where we might display a "loading items...". I can have a play with that and test it on the Pages screen. It might need some design input though.

oandregal commented:
@tellthemachines direction looks good to me, thanks for working on this. I've left some comments about edge cases.

tellthemachines committed:
Fix storybook example

tellthemachines committed:
Fix styling issues with story

tellthemachines committed:
Hide pagination when infinite scroll is enabled

talldan commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -193,6 +195,40 @@ function DataViews< Item >( {
 		}
 	}, [ hasPrimaryOrLockedFilters, isShowingFilter ] );
 
+	const isInfiniteScroll = useMemo( () => {
+		if ( ! infiniteScrollHandler ) {
+			return false;
+		}
+		return !! view.layout?.infiniteScroll;
+	}, [ view, infiniteScrollHandler ] );
```
I think you could remove the `useMemo` here, and instead have the boolean calculated inline.

tellthemachines replied:
We can if we assume that `view.layout?.infiniteScroll` will only exist if the `infiniteScrollHandler` also exists. Which it should! I might have been overly cautious here.

talldan commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -193,6 +195,40 @@ function DataViews< Item >( {
+	}, [ view, infiniteScrollHandler ] );
+
+	// Attach scroll event listener for infinite scroll
+	useEffect( () => {
```
`useRefEffect` could be an option here instead of `useEffect`, but tbh, it's a very minor difference. Only advantage is you wouldn't need to check `containerRef.current` is set.

talldan commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -386,6 +402,7 @@ function ViewGrid< Item >( {
 													config={ {
 														sizes: size,
 													} }
+													posinset={ index + 1 }
```
I wonder if the value should only be passed value when `isInfiniteScroll` is true, because when regular pagination is active it'll be providing the incorrect value (e.g. on the second page it'll start from `1` again).

tellthemachines replied:
Hmm if you're planning to use `posinset` when infinite scroll is disabled, then yes that would make sense to do. Otherwise in this PR `posinset` is only rendered when infinite scroll is on.

talldan replied:
Defo a nitpick. To me, invalid data is passed to GridItem, and then we're trusting that the component does the right thing with that data. It's not an issue with the current code, but code changes, and some future person might miss the subtlety.

tellthemachines replied:
I've found a bit of a problem with the aria attributes on the list layout: they don't play well at all with the `Composite` component we're using. I've removed them from list for now, though I haven't disabled infinite scroll on it. 

`Composit` is built on ariakit and the error occurs on that side. Basically any `posinset` value passed to `Composite.Row` makes it explode 😅

talldan replied:
😬 What's the error?

Might be worth checking the issues for ariakit to see if anyone has had the same - https://github.com/ariakit/ariakit/issues. I searched for posinset and saw nothing though.

talldan replied:
Not sure if you tried it, but you could try setting the prop on the `<div />` that the `Composite.Row` renders.

tellthemachines replied:
`ReferenceError: Cannot access 'rowId' before initialization` which is coming from the ariakit component. I haven't really looked into it. 

Adding it to the div works 🎉  I also had to add `aria-setsize` to the div otherwise it ends up attached to the button inside. 

I also had to update the CSS selectors directly targeting `[role="row"]` to target `[role="article"]` too.

talldan commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -139,6 +143,11 @@ function GridItem< Item >( {
 					);
 				}
 			} }
+			role={ isInfiniteScroll ? 'article' : undefined }
```
Interesting. I think I'll need a different role for the data picker changes. I'm not sure whether `article` supports selection. Will need to figure out if infinite scroll can work for a picker or if that's an invalid combination.

I'll also need `posinset` so it's useful to establish it here.

tellthemachines replied:
That's good question. Aside from whether the roles will work or not, personally I think I'd prefer to infinite scroll through a large picker than to navigate pagination, but it's not a great solution for accessibility so it shouldn't be the default mode. Does it makes sense to have view config options in a picker scenario?

talldan replied:
> Does it makes sense to have view config options in a picker scenario?

Hard to say, possibly a smaller subset of options. But maybe it's not worth differentiating in that case and instead providing all the options.

talldan reviewed (commented):
It's testing well for me so far, other reviewers already noted the same things that I would mention.

Code is looking pretty tidy. I left a few very minor review comments.

tellthemachines committed:
Remove memoed const

tellthemachines committed:
Make posinset optional prop and remove semantics from list layout

tellthemachines committed:
add aria attribs to div wrapper in list layout

tellthemachines committed:
update changelog

andrewserong commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -801,6 +801,7 @@ export function DataviewsViewConfigDropdown() {
 							{ !! activeLayout?.viewConfigOptions && (
 								<activeLayout.viewConfigOptions />
 							) }
+							<InfiniteScrollToggle />
 							<ItemsPerPageControl />
```
Edit: Dan raised a good point that my suggestion below should not be followed!

---

Just thinking out loud, but the toggle control looks quite wide in its current implementation. Would it be worth conditionally wrapping these two controls in an `HStack` (like the sort and direction controls) when infinite scroll is present? (I imagine we'd need to add a `context.hasInfiniteScrollHandler` check here if that's the case?)

Here's how it's looking now:

<img width="402" height="465" alt="image" src="https://github.com/user-attachments/assets/71a8360f-a22e-4214-9479-9824541f11f3" />

But I was wondering if it'd be a bit more efficient use of space if we did something like this:

<img width="422" height="327" alt="image" src="https://github.com/user-attachments/assets/5fd7dc5b-6c86-484b-a807-3db0c94f630b" />

talldan replied:
IMO that makes Items Per Page look too confined. It can also have up to 6 items specified, so while it has 4 in that screenshot it's not a true reflection of the worst case.

andrewserong replied:
Ah, good point! Ignore me then 😄

tellthemachines replied:
Hmm I'm thinking that looks a bit squished 😄 
@jameskoster @jasmussen do you have opinions?

jasmussen replied:
The appearance dropdown works decently as is, and yet it can benefit from a revisit given how it's organically grown. But part of that is also reflecting on: what should be here? IMO, and I feel this somewhat strongly, this dropdown should _not_ hold a toggle for infinite scroll. 

Infinite scroll has pros and cons. If implemented right, for extremely large amounts of data, such as your entire media library archive of tens of thousands, it can be an excellent way to provide you access to the entire silo of content in one view. When paired with other date-scrolling or letter-jumping tools, it can be one of the best ways to find exactly what you're looking for, in a way that maintains performance.

It's also got plenty of downsides. You can't do an in-page search for something that hasn't loaded yet, and there are some assistive tech challenges as well. I recognize those are largely mitigated by the presence of the pagination control, as well as the "Items per page" choice. In most cases, however, those controls are mutually exclusive: either you have infinite scroll, or you have _n_ items per page.

I'm not saying infinite scroll can't be combined with an items per page choice, but it is unusual. And unlike posts per page which is commonly a per-view relevant choice, infinite scroll feels more like a set it and forget it choice; presumably you either want it everywhere, or nowhere. To frame it as an exercise, if we presumed all of WordPress used dataviews for list views, would you set infinite scroll for some, but not other screens? 

For that reason, I'd put infinite scroll as a choice in a global preferences screen, and have it apply to all. Such as Settings > General. 

I recognize that's non-trivial as far as progressing the dataviews component, so I could see it temporarily living in this dropdown, if we were to agree to eventually move it to a global place. 

Although I feel somewhat strongly about this, it's also not a hill I'll do battle on, and if there's consensus otherwise that infinite scroll _should_ be a per-view setting that's tied to or adjacent to posts per page, I would approach it as a toggle button not quite unlike how Figma does it for their auto-layout _wrap_ button:

![wrap](https://github.com/user-attachments/assets/18a56f10-2fe7-47ac-8c00-06c6f4e0637c)

That is, keep the segmented control (or dropdown if need be) for posts per pages, and attach a toggle button to the right of it with a tooltip. This is in part to reduce the prominence of the button.

jameskoster replied:
I agree that the ‘Items per page’ setting loses some meaning when infinite scroll is enabled… my feeling is that we should be more opinionated about how many items to show initially. E.g. calculate and fill the available space, plus a bit beyond. Maybe two or three ‘pages’ worth is a good target? I suppose it’s a bit tricky to calculate, accounting for grid layout, and potentially variable item heights. We may also need to lazy load images (thinking about the Media Library in particular). Hopefully we can come up with a reliable heuristic.

Either way I’d be tempted to hide/ignore the ‘items per page’ setting when infinite scroll is enabled.

tellthemachines replied:
> keep the segmented control (or dropdown if need be) for posts per pages, and attach a toggle button to the right of it with a tooltip

In the interests of getting an initial version in, this might be a good enough approach for now. Should I just use a plain toggle for this? Like:

<img width="51" height="37" alt="Screenshot 2025-08-12 at 11 25 35 am" src="https://github.com/user-attachments/assets/58ad781a-8551-4965-9752-bae75155acc6" />

?

tellthemachines replied:
> Either way I’d be tempted to hide/ignore the ‘items per page’ setting when infinite scroll is enabled

To be clear, I'd also be fine with this 😄 just want to find a compromise for now!

jasmussen replied:
None of these options actually feel good to me. But I nevertheless wanted to visualize what I meant. Curious if it inspires other ideas. What would be the simplest thing to do? And I'm referring here also to the fewest knobs and dials, to help reduce the prominence of the infinite scroll control.

<img width="296" height="446" alt="Frame 1410129757" src="https://github.com/user-attachments/assets/ea607f24-3b1a-4436-9258-d03a37c61f80" />

jameskoster replied:
The toggle buttons feel like mystery meat to me. Infinite scroll isn’t really a behavior that’s easily communicated via icon. Showing the disabled Items per page control feels a bit like wasted space. 

Personally I’d be tempted to show a toggle for infinite scroll, and hide Items per page when infinite scroll is true. This seems the most conventional and acessible solution to me.

<img width="645" height="186" alt="is" src="https://github.com/user-attachments/assets/fb068524-dfdd-477f-b75e-0fb479e6cd5c" />

jasmussen replied:
The two problems I have with the above is that 1) it gives huge prominence to something I'm not sure should have prominence, and secondarily that the panel as a whole is already huge:

<img width="363" height="844" alt="image" src="https://github.com/user-attachments/assets/18391e8e-daa2-4eb7-827f-38fa402486d9" />

The design can probably work if it moves to a Settings > General page eventually. There's space to be verbose and contextualize there. 

I don't want to block this work from moving forward, but the two pieces of followups to look into are just that: moving this to a settings page at some point, and overall looking at whether we can make this list appearance control more dense/compact/glancable. Those two tasks may in fact be the same, I have this vague feeling that most of the "set and forget" settings can live elsewhere.

tellthemachines replied:
> The design can probably work if it moves to a Settings > General page eventually.

One thing we need to be aware of regarding infinite scroll as a global setting is that it has to be implemented on a per-screen basis. Dataviews only provides an event listener that the screen using it can attach a handler to. Even if we add infinite scroll handlers on all the core screens using dataviews, it's up to plugin developers to do it on their screens, so the global setting may only work for a subset of admin screens. Is that acceptable?

tellthemachines replied:
OK I've updated the PR to try out what Jay suggested above:

<img width="353" height="295" alt="Screenshot 2025-08-13 at 3 16 27 pm" src="https://github.com/user-attachments/assets/dc356b83-6eff-4a7a-851a-dee55b0f33bc" />

I've also centered the "loading more results" text:

<img width="857" height="599" alt="Screenshot 2025-08-13 at 3 25 04 pm" src="https://github.com/user-attachments/assets/e9772748-d968-4766-8c66-20fb07a49ad7" />

(that text should only appear over a slow connection)

Does this look OK for now?

jameskoster replied:
For me it’s an acceptable starting point from which to refine. 

I agree the view options dialog is beginning to feel cumbersome in some circumstances, but I see that as a separate issue that we should approach systematically.

A global setting could makes sense, but ultimately this is a property of each individual dataview so the local option would likely need to remain regardless. I could see a global setting working similarly to ‘Default post category’.

jasmussen replied:
Agree we can approach this separately and certainly systematically, and not necessarily urgently.

> A global setting could makes sense, but ultimately this is a property of each individual dataview so the local option would likely need to remain regardless. I could see a global setting working similarly to ‘Default post category’.

Honestly this depends for me. In both Proton Drive, Dropbox, and Google Drive, there's not a choice to turn off or on infinite scroll, or even choose items per page. And it feels plenty, honestly. I recognize that doesn't map 1:1 with what DataViews is doing, because it's also handling column visibility properties. So I'm not suggesting we remove these. Perhaps it's just that the cog feels very prominent, and the dropdown it opens is substantial, so perhaps when we get to it changing the cog to an ellipsis or something ligther, and working on the density of the controls inside, is plenty.

tellthemachines committed:
Add loading more text

tellthemachines committed:
Disable when data by group

tellthemachines commented:
Ok I think that's most of the feedback addressed. 

I added a "loading more items" message when infinite scroll is enabled and data is in the loading state, but it could use some design help:

### table

https://github.com/user-attachments/assets/f5c2b0aa-55f0-4285-ad46-ae08a6064cc2


### grid

<img width="1119" height="567" alt="Screenshot 2025-08-11 at 2 14 32 pm" src="https://github.com/user-attachments/assets/c8e13eea-63ff-4149-85da-c4038a3f65e1" />

### list

<img width="420" height="183" alt="Screenshot 2025-08-11 at 2 30 34 pm" src="https://github.com/user-attachments/assets/899aa073-c069-4b42-b0bb-e5f83e66ce59" />

I've added no styles at all so far, and in table view it appears centred by default. Should we do the same for the other views? 
Also, should we make it a little more visible? Is that font size ok?

andrewserong commented:
> I added a "loading more items" message when infinite scroll is enabled and data is in the loading state, but it could use some design help:

Does it need a loader? I quite liked the look of it without it. Another potential option could be to make the loader configurable by the consumer and default to `null`?

tellthemachines commented:
I added the design feedback label mainly due to https://github.com/WordPress/gutenberg/pull/70955#issuecomment-3173267495 and also https://github.com/WordPress/gutenberg/pull/70955#discussion_r2265639904

oandregal commented on packages/dataviews/src/types.ts:
```diff
@@ -482,11 +482,23 @@ export interface ViewTable extends ViewBase {
 export interface ViewList extends ViewBase {
 	type: 'list';
+
+	layout?: {
```
So far, the `layout` object contained props that were specific to a particular layout and didn't make sense in the others, while infinite scroll seems to fall in the general view config category. Do you see it differently?

tellthemachines replied:
I'm not sure. Currently it applies to all layouts, but will that always be the case? I don't have a strong opinion either way. Where would be a good place for a general setting, just in the view object?

talldan replied:
Or added directly to `ViewBase` rather than the individual ViewList, ViewTable, ViewGrid types?

I feel like that might be what @oandregal was referring to.

tellthemachines replied:
You mean by adding a `layout` object to `ViewBase`? If layout is supposed to be for layout-specific things, I'm not sure we should do that.

talldan replied:
I mean directly at the root of `ViewBase`.

It is optional like most of the other properties so it can still be omitted from the config for particular layouts that don't support it.

oandregal replied:
Yeah, I meant what Dan said: part of the `ViewBase` type and not `layout` :) There is some precedent on props we have in `ViewBase` but we don't know yet how/whether they make sense in _every_ layout (e.g., groupBy, showLevels).

tellthemachines replied:
Ok to be clear, the type being part of `ViewBase` also means that the actual property goes directly in the view object? That's what I was asking initially. Types will naturally be updated as a consequence of that change.

tellthemachines replied:
I've updated to move `infiniteScroll` out of `layout` and directly into `view. This has the side-effect of keeping the state of the toggle when switching between views (whereas before it reset).

tellthemachines commented:
@jasmussen and @jameskoster, thanks for the feedback so far! Can I bother you further to take a look at the screenshots in [this comment](https://github.com/WordPress/gutenberg/pull/70955#issuecomment-3173267495) and let me know whether we're on the right track with the "loading more results..."? Can we ship that with just a few styling adjustments or should we try something different?

tellthemachines committed:
move infiniteScroll to base view

tellthemachines commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -102,12 +103,11 @@ export function ViewTypeMenu() {
 											type: e.target.value,
 											...defaultLayouts[ e.target.value ],
-										} );
+										} as View );
```
Weird that this error disappeared when I added `infiniteScroll` into `layout` in the several views, and reappeared when I moved it out. But I think this solution is nicer than the expect error comment.

tellthemachines committed:
Try simple toggle and hiding per page values

tellthemachines committed:
Center-align "loading more results"

jasmussen commented:
> Can I bother you further to take a look at the screenshots in https://github.com/WordPress/gutenberg/pull/70955#issuecomment-3173267495 and let me know whether we're on the right track with the "loading more results..."? Can we ship that with just a few styling adjustments or should we try something different?

Never a bother.

To specifically answer the question, use a [spinner](https://wordpress.github.io/gutenberg/?path=/docs/components-spinner--docs) instead of the text, which there's every risk you won't be able to read in time. 

As far as the knobs to toggle between items per page and infinite scroll, I've registered my concerns as far as the prominence this receives. As also noted, it's not blocking, if we agree to circle back to this at a later time, for example when we'll be able to move the infinite scroll toggle to a global settings page.

tellthemachines committed:
Change loading more text to spinner

tellthemachines commented:
@jasmussen thanks, I was able to read it in time 😄 

I've now updated to use a spinner on all layout types:

<img width="392" height="249" alt="Screenshot 2025-08-13 at 4 59 03 pm" src="https://github.com/user-attachments/assets/8357b7df-6010-4e30-9719-ebe96d472ab6" />

<img width="1148" height="540" alt="Screenshot 2025-08-13 at 4 59 18 pm" src="https://github.com/user-attachments/assets/6fc0bdfc-4839-47b7-8a9f-811cb70be8bf" />

<img width="1144" height="335" alt="Screenshot 2025-08-13 at 4 59 30 pm" src="https://github.com/user-attachments/assets/1846658e-a0f7-4aa6-b135-fc410b429263" />

tellthemachines committed:
Remove post list handler meant for testing

tellthemachines commented:
Ok peeps this is now ready for a final code review. 

I've removed the post list handler which I'd added just for testing purposes, but you can still test it in storybook: `npm run storybook:dev`

It might be good to look into building a generic infinite scroll handler hook that can be used by several screens, but that can be done as a follow-up at some point.

ramonjd commented on packages/dataviews/src/types.ts:
```diff
@@ -443,6 +443,11 @@ interface ViewBase {
+	/**
+	 * Whether infinite scroll is enabled.
+	 */
+	infiniteScroll?: boolean;
```
Optional: `infiniteScroll` reads more like a noun (the feature itself) rather than a boolean state. I'm wondering if it'd aid readability to have a name that indicates a true/false condition?

Then assignments like `const isInfiniteScroll = view.infiniteScroll;` above would be semantically redundant.

E.g., 


```
infiniteScrollEnabled
hasInfiniteScroll
enableInfiniteScroll
infiniteScrollActive
```

tellthemachines replied:
`infiniteScrollEnabled` is probably the most accurate because that's what the setting checks for.

andrewserong replied:
Oh, good discussion. This is the tiniest of tiny nits, but I see in this PR, there's usage of `infiniteScrollEnabled` and `isInfiniteScrollEnabled`. Would it be good for consistency to double-check that we're naming them the same in each place? I have a mild preference for prefixing with `is` for booleans, but don't feel too strongly about it.

ramonjd commented on packages/dataviews/src/components/dataviews-pagination/index.tsx:
```diff
@@ -22,7 +22,9 @@ export function DataViewsPagination() {
-	if ( ! totalItems || ! totalPages ) {
+	const isInfiniteScrollEnabled = view.infiniteScroll ?? false;
+
+	if ( ! totalItems || ! totalPages || isInfiniteScrollEnabled ) {
```
Idea for a follow up: what do folks think about a go to top navigation? If you've got 100s of images, it might be helpful to go back to the top. I don't know, maybe it's just me 😄 

Another "me" thing is needing an idea of what "infinite" means, e.g., how many items are there.

tellthemachines replied:
Yeah we could have a play with that! I'm not sure how it might work if the handler implementation doesn't just straight up load all items but instead loads next and unloads previous (which it might do for performance reasons, especially when catering to mobile devices). But def something to explore.

ramonjd reviewed (approved):
I tested in storybook (the description still says it can be tested in Pages?)

Works as described for me, and nothing stands out in the code. 

If the design folks are happy with it, then LGTM too for what it's worth, and it can be iterated upon before adding it to the post list as well.

Just left a suggestion re: the boolean var name.

Good show!

tellthemachines committed:
Update property name and clean up

andrewserong reviewed (approved):
This is looking great to me, too. I really like the API of `infiniteScrollEnabled` in the view, and `infiniteScrollHandler` in the pagination info — it feels like a good starting point to me and will allow consumers a lot of flexibility to explore a variety of different implementations.

LGTM! 🚀

tellthemachines commented:
Thanks for the reviews everyone! 🎉
